### PR TITLE
Performance optimization

### PR DESCRIPTION
**Pull request by Kaidesuyoo**
Hello VSCode team.
I've made some performance optimizations and the most impactful change is here:
```typescript
// before
const cancellationTokenListener = cancellationToken.onCancellationRequested(cancel);
disposable = combinedDisposable(toDisposable(cancel), cancellationTokenListener);

//after
disposable = cancellationToken.onCancellationRequested(cancel);
``` 
## Why?
Before the modification, the `cancel` function was added to the disposable. In my understanding, this `cancel` function should be used to promptly cancel the request when a cancellation occurs.
However, when the request has not been canceled and a response is received, the result promise reaches the finally phase:
```typescript
return result.finally(() => {
	// This will result in the cancel function being executed regardless.
	disposable.dispose();
	this.activeRequests.delete(disposable);
});
``` 
This is what the cancel function does:
```typescript
const cancel = () => {
	if (uninitializedPromise) {
		uninitializedPromise.cancel();
		uninitializedPromise = null;
	} else {
		// After the client connects, a cancellation request is sent after each request is completed.
		this.sendRequest({ id, type: RequestType.PromiseCancel });
	}
	e(new CancellationError());
};
``` 
On the server side, the received requests are cleaned up after processing, meaning that each cancellation request triggered by the client results in an unnecessary cleanup on the [server](https://github.com/microsoft/vscode/blob/main/src/vs/base/parts/ipc/common/ipc.ts#L444).
```typescript
// server side
finally(() => {
	disposable.dispose();
	// Delete after processing the request.
	this.activeRequests.delete(request.id);
});

// After the server returns the processed request, it receives a cancellation request from the client.
private disposeActiveRequest(request: IRawRequest): void {
	const disposable = this.activeRequests.get(request.id);
	// The request has already been processed and deleted, so the following code will not execute.
	if (disposable) {
		disposable.dispose();
		this.activeRequests.delete(request.id);
	}
}
``` 
In summary, I believe that the cancellation requests here are meaningless and occur almost for every request established through IPC communication.

This situation started occurring after [this](https://github.com/microsoft/vscode/commit/39de20e8a3b4aa1f59f194a97f69ae2e3efa2b73#diff-ecb6cabcd3ef41af04520359965058c04dd30e9712d8a4e22f60fd23d0da3830R643) change.

### CODE REVIEW

**Comment by Kaidesuyoo** [comment]
@microsoft-github-policy-service agree

File: src/vs/base/parts/ipc/common/ipc.ts [hunk 1]
```diff
@@ -655,12 +654,19 @@ export class ChannelClient implements IChannelClient, IDisposable {
...
+					doRequest();
+				} else {
+					uninitializedPromise = createCancelablePromise(_ => this.whenInitialized());
+					uninitializedPromise.then(() => {
+						uninitializedPromise = null;
+						doRequest();
+					});
+				}
 			},
 			onDidRemoveLastListener: () => {
```
Comment (connor4312) [inline src/vs/base/parts/ipc/common/ipc.ts:671]: This block will no longer remove the listener if state is idle when the event is listened to

**Comment by connor4312** [comment]
I'm not sure how your PR description relates to the code you changed. Your code is in `requestEvent` but you refer to processing in `requestPromise`.

Cancellation is also not requested when disposing a CancellationTokenSource by default so I'm not sure that I see where the problem is occurring -- were you able to verify what you say actually happens when debugging server?

File: src/vs/base/parts/ipc/common/ipc.ts (hunk 1 above)
Comment (Kaidesuyoo) [reply src/vs/base/parts/ipc/common/ipc.ts:671]: Regarding your suggestion, I have a question: 
when the state is idle, it indicates that `uninitializedPromise` has completed execution and has been set to `null`. Therefore, when `onDidRemoveLastListener` occurs, it should correctly trigger `EventDispose`. Is there something I might have overlooked?

**Comment by Kaidesuyoo** [comment]
> I'm not sure how your PR description relates to the code you changed. Your code is in `requestEvent` but you refer to processing in `requestPromise`.
> 
> Cancellation is not requested when disposing a CancellationTokenSource by default so I'm not sure that I see where the problem is occurring -- were you able to verify what you say actually happens when debugging the server?

I apologize for the confusion. The code in requestEvent is a different part; what I described in the PR pertains to the code within requestPromise.

As you mentioned, disposing of the `CancellationTokenSource` does not trigger the cancellation request. The issue I'm referring to is that the `cancel` function is added to the `disposable`, which causes the `cancel` function to be triggered upon disposal.

Here’s what happened during debugging:

https://github.com/user-attachments/assets/527c4416-6809-4557-bf39-b00cee315d04

File: src/vs/base/parts/ipc/common/ipc.ts (hunk 1 above)
Comment (connor4312) [reply src/vs/base/parts/ipc/common/ipc.ts:671]: Oh, nevermind, that should be fine

File: src/vs/base/parts/ipc/common/ipc.ts [hunk 2]
```diff
@@ -635,8 +635,7 @@ export class ChannelClient implements IChannelClient, IDisposable {
 				e(new CancellationError());
 			};
 
-			const cancellationTokenListener = cancellationToken.onCancellationRequested(cancel);
-			disposable = combinedDisposable(toDisposable(cancel), cancellationTokenListener);
```
Comment (connor4312) [inline src/vs/base/parts/ipc/common/ipc.ts:639]: I think we do still want to cancel the request and reject the promise when the disposable is disposed, but only if it the promise already been resolved. I.e. it should so as it was doing if the client is disposed with ongoing requests (otherwise those requests just hang indefinitely) but not from the `result.finally` block.

Comment (Kaidesuyoo) [reply src/vs/base/parts/ipc/common/ipc.ts:639]: You're right. 
I made some changes, could you take another look?

File: src/vs/base/parts/ipc/common/ipc.ts [hunk 3]
```diff
@@ -635,14 +637,20 @@ export class ChannelClient implements IChannelClient, IDisposable {
...
-			this.activeRequests.add(disposable);
+			disposable = combinedDisposable(cancellationToken.onCancellationRequested(cancel));
```
Comment (connor4312) [inline src/vs/base/parts/ipc/common/ipc.ts:640]:
```suggestion
			disposable = cancellationToken.onCancellationRequested(cancel);
```

**Review by connor4312 — APPROVED** [review]

**Comment by connor4312** [comment]
Thanks for the PR!

**Review by Tyriar — APPROVED** [review]
